File: commands/wallet.py
import requests

RPC_URL = "https://eth-sepolia.blockscout.com/api/eth-rpc"
API_URL = "https://eth-sepolia.blockscout.com/api/v2"
HEADERS = {"Content-Type": "application/json"}

def eth_rpc(method, params):
    payload = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": method,
        "params": params
    }
    resp = requests.post(RPC_URL, json=payload, headers=HEADERS)
    resp.raise_for_status()
    return resp.json().get("result")

def format_number(num):
    if isinstance(num, (int, float)):
        return f"{num:,}"
    return num

def cmd_wallet(args):
    if not args:
        print("Usage: wallet [address]")
        return

    address = args[0]
    print(f"\nWallet Information for: {address}\n")

    try:
        balance_wei = int(eth_rpc("eth_getBalance", [address, "latest"]), 16)
        balance_eth = balance_wei / 10**18

        tx_count = int(eth_rpc("eth_getTransactionCount", [address, "latest"]), 16)

        code = eth_rpc("eth_getCode", [address, "latest"])
        is_contract = code != "0x"

        response = requests.get(f"{API_URL}/addresses/{address}")
        response.raise_for_status()
        address_data = response.json()

        token_response = requests.get(f"{API_URL}/addresses/{address}/tokens")
        token_response.raise_for_status()
        token_data = token_response.json()

        # Extract information from the API responses
        last_balance_block = address_data.get("block_number_balance_updated_at", "N/A")
        has_token_transfers = address_data.get("has_token_transfers", False)
        tokens_count = len(token_data.get("items", []))
        proxy_implementation = "N/A"

        if address_data.get("implementations"):
            implementations = address_data.get("implementations", [])
            if implementations and len(implementations) > 0:
                proxy_implementation = implementations[0].get("address", "Unknown")

        # Get transaction data to calculate gas used
        tx_list_response = requests.get(f"{API_URL}/addresses/{address}/transactions")
        tx_list_response.raise_for_status()
        tx_list_data = tx_list_response.json()

        total_gas_used = 0
        transfers_count = 0

        for tx in tx_list_data.get("items", []):
            if "gas_used" in tx:
                total_gas_used += int(tx.get("gas_used", 0))
            if tx.get("method") == "transfer":
                transfers_count += 1

        # Format output
        print(f"Delegated to:      {proxy_implementation if is_contract else 'N/A'}")
        print(f"Balance:           {balance_eth:.8f} ETH")
        print(f"Tokens:            {tokens_count}")
        print(f"Net worth:         N/A")
        print(f"Transactions:      {tx_count}")
        print(f"Transfers:         {transfers_count if has_token_transfers else 0}")
        print(f"Gas used:          {format_number(total_gas_used)}")
        print(f"Last balance update: {last_balance_block}")
        print(f"Account Type:      {'Contract' if is_contract else 'EOA (External Owned Account)'}")

    except Exception as e:
        print(f"Error retrieving wallet info: {e}")

File: typescript_logic/wallet.ts
import * as axios from "axios";

const RPC_URL = "https://eth-sepolia.blockscout.com/api/eth-rpc";
const API_URL = "https://eth-sepolia.blockscout.com/api/v2";
const HEADERS = { "Content-Type": "application/json" };

async function ethRpc(method: string, params: any[]): Promise<any> {
    const payload = {
        jsonrpc: "2.0",
        id: 1,
        method: method,
        params: params
    };

    try {
        const response = await axios.post(RPC_URL, payload, { headers: HEADERS });
        return response.data.result;
    } catch (error) {
        console.error(`Error in ethRpc: ${error}`);
        throw error;
    }
}

function formatNumber(num: number | string): string {
    if (typeof num === "number") {
        return num.toLocaleString();
    }
    return num;
}

export async function cmdWallet(args: string[]): Promise<void> {
    if (args.length === 0) {
        console.log("Usage: wallet [address]");
        return;
    }

    const address = args[0];
    console.log(`\nWallet Information for: ${address}\n`);

    try {
        // Get balance in Wei and convert to ETH
        const balanceWei = parseInt(await ethRpc("eth_getBalance", [address, "latest"]), 16);
        const balanceEth = balanceWei / 10 ** 18;

        // Get transaction count
        const txCount = parseInt(await ethRpc("eth_getTransactionCount", [address, "latest"]), 16);

        // Check if the address is a contract
        const code = await ethRpc("eth_getCode", [address, "latest"]);
        const isContract = code !== "0x";

        // Fetch address data from the API
        const addressResponse = await axios.get(`${API_URL}/addresses/${address}`);
        const addressData = addressResponse.data;

        // Fetch token data
        const tokenResponse = await axios.get(`${API_URL}/addresses/${address}/tokens`);
        const tokenData = tokenResponse.data;

        // Extract information from the API responses
        const lastBalanceBlock = addressData.block_number_balance_updated_at || "N/A";
        const hasTokenTransfers = addressData.has_token_transfers || false;
        const tokensCount = tokenData.items?.length || 0;
        let proxyImplementation = "N/A";

        if (addressData.implementations && addressData.implementations.length > 0) {
            proxyImplementation = addressData.implementations[0].address || "Unknown";
        }

        // Fetch transaction data to calculate gas used
        const txListResponse = await axios.get(`${API_URL}/addresses/${address}/transactions`);
        const txListData = txListResponse.data;

        let totalGasUsed = 0;
        let transfersCount = 0;

        for (const tx of txListData.items || []) {
            if (tx.gas_used) {
                totalGasUsed += parseInt(tx.gas_used, 10);
            }
            if (tx.method === "transfer") {
                transfersCount += 1;
            }
        }

        // Format output
        console.log(`Delegated to:      ${isContract ? proxyImplementation : "N/A"}`);
        console.log(`Balance:           ${balanceEth.toFixed(8)} ETH`);
        console.log(`Tokens:            ${tokensCount}`);
        console.log(`Net worth:         N/A`);
        console.log(`Transactions:      ${txCount}`);
        console.log(`Transfers:         ${hasTokenTransfers ? transfersCount : 0}`);
        console.log(`Gas used:          ${formatNumber(totalGasUsed)}`);
        console.log(`Last balance update: ${lastBalanceBlock}`);
        console.log(`Account Type:      ${isContract ? "Contract" : "EOA (Externally Owned Account)"}`);
    } catch (error) {
        console.error(`Error retrieving wallet info: ${error}`);
    }
}